README.md
# general.js

**JS Library to improve your web**
 
JS Library that handles DOM / Events - DOM / AJAX - FETCH / WebSockets-Webworkers / Watch - UnWatch / Crypto / expansible 

## How to init the library
```javascript
	genrl.docready(function(){
		//write code below
	});
```

## Log something to JS console
```javascript
	genrl.log("Hello World");
```

## How to select a DOMElement By ID
```javascript
	g("#element");
```

## How to select a DOMElement By ClassName
```javascript
	g(".element");
```

## How to read an object
```javascript
genrl.each(h,function(x,index){
	genrl.log("RESULTADO");
	genrl.log(x.nombre);
	genrl.log(index);
})
```

## How to set CSS attributes
```javascript
	g("#element").css({
		'height': '400px',
		'width': '200px',
		'opacity': '1'
	});
```
## Bind an event
```javascript
g("#btnmover").click(function(){
	g("#div_A").animate('bounce',5000,function(){
		genrl.log("Se ha activado el callBack");
	});
});
```
## How to get an attribute of a DOMElement
```javascript
	idelement=g("#element").getAttrb("id");
```

## How to set an attribute of a DOMElement
```javascript
	g("#element").setAttrb("attribRandom","test");
```

## Get element children
```javascript
	g("#element").children();
```

## Get element child number N
```javascript
	g("#element").child(N);
```

## ROUTES
```javascript
	genrl.path.map("#/prueba").to(function(){
		g("#cargadiv").load("README.md",function(){
			genrl.log("Módulo cargado.");
		});
	});
	genrl.path.listen();
```

## Get fileList from file input when it changes
### archivo is a file input
```javascript
	g('#archivo').change(function(e){
		console.log("Cambió el campo");
		dataf=g('#archivo').getFiles();
	});
```

## AJAX Calls
### GET + Callback
```javascript
	let fetchobj=genrl.ajaxapi;
	g("#getbtn").click(function(){
		fetchobj
		.get("general.js/README.md")
		.then(function(data){
			console.log("DATA: " + data);
			g("#titulo_widget").html("RESULTADO:");
			g("#mensajes").html(data);
		})
		.catch(function(e){	
			console.log("ERROR:" + e);
		})
	});
```
### GET JSON + Callback
```javascript
	let fetchobj=genrl.ajaxapi;
	g("#getjbtn").click(function(){
		fetchobj
		.getJSON("http://localhost/sistemapmod/devtools/dev/general.js/config.json")
		.then(function(data){
			console.log("DATA: " + data);
			g("#titulo_widget").html("RESULTADO:");
			g("#mensajesb").html(data);
		})
		.catch(function(e){	
			console.log("ERROR:" + e);
		})
	});
```
## Load asinchronous + Callback
```javascript
	let fetchobj=genrl.ajaxapi;
	g("#loadbtn").click(function(){
		fetchobj
		.load("http://localhost/sistemapmod/devtools/dev/general.js/README.md")
		.then(function(data){
			console.log("DATA: " + data);
			g("#titulo_widget").html("RESULTADO:");
			g("#mensajesa").html(data);
		})
		.catch(function(e){	
			console.log("ERROR:" + e);
		})
	});
```

### Asynchronous POST for General.JS 
#### Client Side
```javascript
	let fetchobj=genrl.ajaxapi;
	g("#namebtn").click(function(){
		let strdata={'nombre':'arturo'};
		datos=strdata;
		fetchobj
		.post("socketd.php",datos)
		.then(function(data){
			console.log("DATA RECIBIDA: ");
			console.log(data);
			g("#titulo_widget").html("RESULTADO:");
			g("#mensajesa").html(data);
		})
		.catch(function(e){	
			console.log("ERROR:" + e);
		})
	});
```

### Asynchronous Files UPLOAD for General.JS 
#### Client Side
```javascript
	let fetchobj=genrl.ajaxapi;
	g('#archivo').change(function(e){
		console.log("Cambió el campo");
		dataf=g('#archivo').getFiles();
	});
	g("#filebtn").click(function(){
		let fdata = new FormData();
		fdata.append("file", dataf[0]);
		fetchobj
		.upload("uploadfile.php",fdata)
		.then(function(data){
			g("#titulo_widget").html("RESULTADO:");
			g("#mensajesb").html(data);
		})
		.catch(function(e){	
			console.log("ERROR:" + e);
		})
	});
```
#### Server Side
```php
<?php
	session_start();
    if (move_uploaded_file($_FILES["file"]["tmp_name"], "uploads/".$_FILES['file']['name'])) {
        //more code here...
        echo("uploads/".$_FILES['file']['name']);
    }
?>
```

## Smooth scrolling
```javascript
	g("#holap").click(function(){
		g("#holap").smooth("#adiosp",{
			duration:'10000',
			offset: 0,
			callback: function(){
				g.log("Scroll finalizado");
			}
		});
	});
```

### Asynchronous JS FETCH API
```javascript
	genrl.log("*****FETCH API*****");
	var fapi=genrl.fetchapi.getFetch();
	if(fetchapi){
		g.log("FETCH API ha sido cargada exitosamente!");
		g.log("***************************************");
		var dataSrc={};
		dataSrc={
			method:'fetch',
			name:"arturo",
			lastname:"vasquez"
		}
		genrl.fetchapi.post(
			"socket_fetch.php",
			dataSrc,
			function(data){
				genrl.log("DATA FETCH");
				genrl.log(data);
			}
		);
		genrl.fetchapi.get(
			"example.json",
			function(data){
				var datappal;
				var datos;
				genrl.log("DATA FETCH GET");
				genrl.log(data);
				datappal=data.data;
				genrl.log("PERSONA");
				genrl.log(datappal.persona);
				var datos=datappal.persona;
				genrl.log(datos.ciudad);
				genrl.log("**************");
				genrl.log(datos.nombre);
				genrl.log("**************");
				genrl.log(datos.apellido);
			}
		);
	}
```

## How to extend the library
### Creates a new function
```javascript
	genrl.extend({
		myfunction:function(options){
			//write code below
		});
	});
```

### Use new function 
```javascript
	genrl.myfunction(options);
```

Use the Source...

The Source be with you...


CHANGELOG.md
# Change Log
- All notable changes to this project will be documented in this file.
 
- The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).
 
## [0.0.1] - 2021-04-03
 
- Added AJAX API
 
### Added
- AJAXAPI
-   MINOR changes added UPLOAD and getJSON and LOAD
  TO have GET, POST, getJSON, UPLOAD
 
### Changed
  
- FETCHAPI
-  PATCH Fetch module revealed pattern
 
### Fixed

- AJAX API Get Function
- AJAX API GetJSON Function
- AJAX API Load Function
- AJAX API Post Function
- AJAX API Upload Function
- Refact AJAX API functions


app.js
genrl.docready(function(){
	var cadenamd5;
	var h={};
	var contdivs;
	var bitvisible;
	var cadena;
	var elementos=[];
	var ww={};
	var index=0;

	h=[
		{'nombre':'arturo'},
		{'nombre':'pedro'},
		{'nombre':'antonio'},
		{'nombre':'oscar'},
		{'nombre':'nepo'},
		];
	genrl.log("**************prop Function************");
	//buscar sin ID pero sabiendo la etiqueta
	//genrl.log(g.dom("html").prop("data-scope"));
	//buscar con ID, sabiendo el atributo
	genrl.log(genrl.ds("#appdata0").get("scope"));
	//buscar solo el atributo
	elementos=genrl.propAll(".abc");
	genrl.log("**************ELEMENTOS************");
	genrl.log(elementos);
	genrl.log("**************ELEMENTOS************");
	genrl.log("**************prop Function************");
	genrl.log("H ARRAY");
	genrl.each(h,function(x,index){
		genrl.log("RESULTADO");
		genrl.log(x.nombre);
		genrl.log(index);
	})
	genrl.log("FIN H ARRAY");
	nav=genrl.browser();
	genrl.log("**********NAV**************");
	genrl.log(nav);
	genrl.log("**********NAV**************");
	cadenamd5=genrl.cripto.encode("MD5","pruebaclave","pruebaclave");
	genrl.log("CLAVE MD5: " + cadenamd5);
	g("#databind").on('keyup',function(e){
		var valornew;
		valornew=g("#databind").val();
		g("#dataresp").val(valornew);
	});
	contdivs=0;
	bitvisible=0;

	g("#btnmover").click(function(){
		g("#div_A").animate('bounce',1500,function(){
			genrl.log("Se ha activado el callBack");
		});
	});
	g("#btnmostrar").click(function(){
		g("#ajaxresponse").animate('fadeIn',5000,function(){
			genrl.log("Finished");
		});
	});
	g("#btnocultar").click(function(){
		g("#ajaxresponse").animate('fadeOut',5000,function(){
			genrl.log("Finished");
		});
	});
	///////////////////////////////////////
	genrl.log("Path JS Version: " + genrl.path.getVersion());
	genrl.log("****************************FUNCION DATA:****************************");
	g("div").addAttrb("gn-repeat","none");
	g(".abc").addAttrb("gn-repeat","none");
	g(".abc").data("attrib-nuevo","true");
	result=g(".abc").getAttrb("gn-repeat");
	resultdos=g(".abc").data("attrib-nuevo");
	genrl.log(resultdos[0]);
	genrl.log("******************************************************************");
	g(".abc").rmAttrb("gn-repeat");
	//scrollLeft - Getter/Setter
	g("#contentCont").scrollLeft();
	g("#contentCont").scrollLeft(200);
	//scrollTop
	g("#contentCont").scrollTop();
	g("#contentCont").scrollTop(100);
	g("#holap").cursor('pointer');
	//Por arreglo simple de un solo elemento
	g("#holap").css({'opacity':'1'});
	g("#holap").css({'color':'#000'});
	//en JSON
	g("#holap").css({
		'position':'relative',
		'display':'block',
		'float':'none',
		'background':"#fff000",
		'font-size':"20px",
		'margin-top':'20px',
		'margin-bottom':'20px',
		'border':'1px solid #000',
		'text-align':'center',
		'content':'Esto es un titulo'
	},function(){
		genrl.log("cambios de estilo realizados");
	});
	genrl.log("HEIGHT****************************");
	var props=g(".abc").css("height");
	genrl.log(props);
	genrl.log("HEIGHT****************************");
	genrl.log("HEIGHT****************************");
	var props=g(".abc").css(["height","width"]);
	genrl.log("OBJFINAL****************************");
	genrl.log(props);
	genrl.log("OBJFINAL****************************");
	genrl.log("HEIGHT****************************");
	genrl.log("***ARRAY HEIGHT****************************");
	g("#holap").click(function(){
		g("#holap").smooth("#adiosp",{
			duration:'10000',
			offset: 0,
			callback: function(){
				genrl.log("Scroll finalizado");
			}
		});
	});
	g("#adiosp").click(function(){
		g("#holap").smooth("#adiosp",{
			duration:'10000',
			offset: 0,
			callback: function(){
				genrl.log("Scroll finalizado");
			}
		});
	});
	g("#btnfile").click(function(){
		var archivo;
		//holaf=id del control file
		//function(data)=callback con return de la data obtenida
		g.upload("#holaf",function(data){
			//tratar variable para convertir string en JSON
			//Imprimir variable///////////////////////////////
			genrl.log("*******************data.file*******************");
			genrl.log("DATA");
			genrl.log(data);
			genrl.log("DATA FILE NAME");
			genrl.log(data.file);
			genrl.log("DATA FILE STATUS");
			genrl.log(data.status);
			genrl.log("*******************data.file*******************");
			//////////////////////////////////////////////////
		});
	});
	g("#btnwebsock").click(function(){
		genrl.log("Mensaje a enviar " + g("#mensaje").val());
		genrl.ws.reply('socket1',g("#mensaje").val(),function(){
			g("#responsechat").append(g("#mensaje").val());
		});
	});
	genrl.getJSON(
		{
			varu:"arturo",
			vard:"vasquez"
		},
		"socket.php",
		function(data){
			genrl.log("DATA: ");
			genrl.log(data.data);
		}
	);
	genrl.post(
		{
			varu:"arturo",
			vard:"vasquez"
		},
		"socket.php",
		function(data){
			genrl.log("DATA: ");
			genrl.log(data.data);
			g("#ajaxresponse").html("RESPUESTA AJAX \n" + data.data);
		}
	);
	//Extender el objeto con otro objeto llamado saludo
	saludo=(function(){
		return{
			hola:function(mensaje){
				genrl.log("HOLA, " + mensaje);
			}
		}
	}());
	genrl.extend({
		saludo
	});
	saludofunc=(function(){
		genrl.log("HOLA, " + mensaje);
	}());
	genrl.extend({
		saludofunc
	});
	genrl.log(genrl);
	genrl.saludo.hola("Arturo -> funcion extendida con otro objeto");
	//Extender el objeto con funcion llamado saludofunc
	genrl.saludofunc("Arturo 1");
	//Extender el objeto con funcion llamado saludos en fn
	genrl.fn.saludos("Arturo 2");
	genrl.storage.setLocal("usuario","Arturo");
	usuariop=genrl.storage.getLocal("usuario");
	genrl.log("USUARIO " + usuariop);
	genrl.log("*****Bit Verif isReady*****");
	genrl.log(g.isReady());
	genrl.log("*****Bit Verif isReady*****");
	genrl.log("*****WebWorker 1*****");
	genrl.ww.set('worker1',"./src/mods/worker1.js");
	genrl.ww.send('worker1',"Hola");
	genrl.log("*****WebSocket 1*****");
	genrl.ws.set('socket1',"ws://127.0.0.1:5001");
	genrl.log("*****WebSocket 1*****");
	genrl.path.map("#/prueba").to(function(){
		g("#cargadiv").load("README.md",function(){
			genrl.log("Módulo cargado.");
		});
	});
	genrl.path.listen();
	genrl.log("*****FETCH API*****");
	var fapi=genrl.fetchapi.getFetch();
	if(fetchapi){
		genrl.log("FETCH API ha sido cargada exitosamente!");
		genrl.log("***************************************");
		var dataSrc={};
		dataSrc={
			method:'fetch',
			name:"arturo",
			lastname:"vasquez"
		}
		genrl.fetchapi.post(
			"socket_fetch.php",
			dataSrc,
			function(data){
				genrl.log("DATA FETCH");
				genrl.log(data);
			}
		);
		genrl.fetchapi.get(
			"example.json",
			function(data){
				var datappal;
				var datos;
				genrl.log("DATA FETCH GET");
				genrl.log(data);
				datappal=data.data;
				genrl.log("PERSONA");
				genrl.log(datappal.persona);
				var datos=datappal.persona;
				genrl.log(datos.ciudad);
				genrl.log("**************");
				genrl.log(datos.nombre);
				genrl.log("**************");
				genrl.log(datos.apellido);
			}
		);
	}
	//plugin GcycleJS para GeneralJS////
	genrl.log("GCYCLE SLIDER");
	var slider=gcycle.run({
		container:"#slideshowslide",
		slidesToShow:1,
		dots:".dots",
		draggable:true,
		duration:1,
		rewind:true,
		autoplay:true,
		durationbs:3000,
		arrows:{
			prev:'.glider-prev',
			next:'.glider-next',
		},
		getobject:true
	});
	genrl.log("**********************************");
	genrl.log("**********************************");
	genrl.log("MODAL API");
	g("#btnmodal").click(function(){
		var modalwin=genrl.modalapi.getModal({
			title:"Hola",
			html:true,
			close:true,
			content:"Este es el contenido"
		});
		modalwin.open();
	});
	genrl.log("**********************************");
	genrl.log("**********************************");
	genrl.log("**********************************");
	genrl.log("VIDEO API");
	var video=genrl.ytapi.play({
		'cont':'#player_out',
		'id':'WyOIPZ2hLSA',
		'ancho':'1024',
		'alto':'1024'
	});
	genrl.log(video);
	genrl.log("**********************************");
});
